File: tasks/2026/round2_pp/airplane-pp.py
#!/usr/bin/env python3

import sys


INF = 2_000_000_000


def main():
    data = list(map(int, sys.stdin.read().split()))
    it = iter(data)

    T = next(it)
    out_lines = []

    for _ in range(T):
        n = next(it)
        h = next(it)
        w = next(it)

        xs = [next(it) for _ in range(w)]
        ys1 = [next(it) for _ in range(w)]
        ys2 = [next(it) for _ in range(w)]

        # Build and sort walls, add sentinel
        walls = [(xs[i], ys1[i], ys2[i]) for i in range(w)]
        walls.append((INF, 0, 0))
        walls.sort(key=lambda t: t[0])
        wi = 0

        # DP arrays: min and max vertical speed (angle) at each cell
        min_angle = [[INF] * h for _ in range(n)]
        max_angle = [[-INF] * h for _ in range(n)]

        sy = h // 2
        min_angle[0][sy] = 0
        max_angle[0][sy] = 0

        for x in range(n - 1):
            wx, wy1, wy2 = walls[wi]
            is_wall_col = (x == wx)

            for y in range(h):
                if is_wall_col and (y < wy1 or y > wy2):
                    # This cell is blocked by a wall (no hole here)
                    continue

                vmin = min_angle[x][y]
                vmax = max_angle[x][y]
                if vmin > vmax:
                    # This cell is unreachable
                    continue

                next_x = x + 1
                row_min = min_angle[next_x]
                row_max = max_angle[next_x]

                # Try all possible new speeds: current range +/- 1
                for v in range(vmin - 1, vmax + 2):
                    ny = y + v
                    if 0 <= ny < h:
                        if v < row_min[ny]:
                            row_min[ny] = v
                        if v > row_max[ny]:
                            row_max[ny] = v

            if is_wall_col:
                wi += 1

        last_col_min = min_angle[n - 1]
        last_col_max = max_angle[n - 1]
        possible = any(last_col_min[y] <= last_col_max[y] for y in range(h))
        out_lines.append("YES" if possible else "NO")

    sys.stdout.write("\n".join(out_lines))


if __name__ == "__main__":
    main()

File: tasks/2026/round2_pp/airplane-pp.cpp
// @check-accepted: examples small_h medium_h_one_wall medium_h_two_walls medium_h one_wall two_walls no-limits
// insert brief description of the solution here

#include <fstream>
#include <iostream>
#include <vector>
#include <algorithm>

struct Wall {
    int x;
    int y1, y2;
};

using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(0);


    int T;
    cin >> T;
    for (; T > 0; T--) {
        int n, h, w;
        cin >> n >> h >> w;
        vector<Wall> walls(w + 1);
        walls[w].x = 2'000'000'000;
        for (int i = 0; i < w; i++) {
            cin >> walls[i].x;
        }
        for (int i = 0; i < w; i++) {
            cin >> walls[i].y1;
        }
        for (int i = 0; i < w; i++) {
            cin >> walls[i].y2;
        }

        sort(walls.begin(), walls.end(), [](Wall a, Wall b){return a.x < b.x;});
        int wi = 0;

        vector<vector<int>> min_angle(n, vector<int>(h, 2'000'000'000));
        vector<vector<int>> max_angle(n, vector<int>(h, -2'000'000'000));

        int sy = h / 2;
        min_angle[0][sy] = max_angle[0][sy] = 0;

        for (int i = 0; i < n - 1; i++) {
            for (int j = 0; j < h; j++) {
                if (i == walls[wi].x) {
                    if (j < walls[wi].y1 || j > walls[wi].y2) continue;                
                }
                for (int k = min_angle[i][j] - 1; k <= max_angle[i][j] + 1; k++) {
                    int ny = j + k;
                    if (ny < 0) continue;
                    if (ny >= h) continue;
                    
                    int nx = i + 1;
                    min_angle[nx][ny] = min(min_angle[nx][ny], k);
                    max_angle[nx][ny] = max(max_angle[nx][ny], k);
                }            
            }
            if (i == walls[wi].x) {
                wi++;        
            }
        }

        bool possible = false;
        for (int j = 0; j < h; j++) {
            if (min_angle[n - 1][j] <= max_angle[n - 1][j]) {
                possible = true;        
            }    
        }
        if (possible) {
            cout << "YES\n";
        }
        else {
            cout << "NO\n";
        }
    }

    return 0;
}
